nupic/research/frameworks/pytorch/hooks/hook_manager.py

```diff
@@ -0,0 +1,169 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from nupic.research.frameworks.pytorch.model_utils import filter_modules
+
+from .base import TrackStatsHookBase
+
+
+class ModelHookManager:
+    """
+    This class registers and manages a set of hooks of subclassed from
+    `TrackStatsHookBase`. The given hook is registered on all modules matching the given
+    `include_*` criteria. By default, if no criteria is specified, all modules will be
+    tracked. See `filter_modules`_ for further details on the `include_*` params.
+
+    Tracking is started and stopped for all hooks via `self.start_tracking()` and
+    `self.stop_tracking()`. Alternatively, this class can be used a context manager to
+    automate these calls. For example,
+
+    ```
+    with hook_manager as hooks:
+        ... # Train here
+         stats = hooks.get_statitics()
+    ```
+
+    .. _filter_modules: nupic.research.frameworks.pytorch.model_utils.filter_modules
+
+    :param model: model to add hooks to
+    :param hook_class: class subclassed from `TrackStatsHookBase`
+    :param hook_type: whether to register the hook as "forward" or "backward"
+    :param include_modules: a list of module types to track
+    :param include_names: a list of module names to track e.g. "features.stem"
+    :param include_patterns: a list of regex patterns to compare to the names; for
+                             instance, all feature parameters in ResNet can be included
+                             through "features.*"
+    """
+
+    def __init__(
+        self, model, hook_class,
+        hook_type="forward",
+        include_modules=None,
+        include_names=None,
+        include_patterns=None,
+    ):
+
+        assert hook_type in ["forward", "backward"]
```

Comment:
No pre_forward?

Reply:
Fair point, I can add it. I also buffed up the documentation for `TrackStatsHookBase` to reflect the full range of hooks it can be used for.